src/OptimalFilter.jl

```diff
@@ -0,0 +1,512 @@
+                                stations::StationVectors, params::TestParameters) where T
+
+    # The arguments for the WΛWH decompositions are matrices that only have nonzero values
+    # at the indices of the stations. Store them as sparse arrays to save space.
+    mu21 = sparse(stations.ist, stations.jst,
+                  offline_matrices.R22_inv * (offline_matrices.mu20 * observations),
+                  params.nx+1, params.ny+1)
+    mu22 = params.sigma_obs^(-2) * sparse(stations.ist, stations.jst,
+                                          observations,
+                                          params.nx+1, params.ny+1)
+
+    # Compute WΛWH decompositions, results are dense matrices, store them in buffers
+    # These correspond to mu21 and mu22 in Alex's code
+    WΛWH_decomposition!(offline_matrices.buf1, mu21, offline_matrices, params)
+    WΛWH_decomposition!(offline_matrices.buf2, mu22, offline_matrices, params)
+
+    # Compute the difference of the decomposition results, store in offline_matrices.buf1
+    # This corresponds to mu2 in Alex's code.
+    # TODO: Check if WΛWH is linear and you could swap the order
+    offline_matrices.buf2 .-= offline_matrices.buf1
+
+    mu10 = Vector{T}(undef, params.nobs)
+
+    # Loop over particles
+    for iprt = 1:params.nprt
+
+        mul!(mu10, offline_matrices.R22_inv, get_values_at_stations(@view(height[iprt,:,:]), stations))
+        mu10_sparse = sparse(stations.ist, stations.jst, mu10, params.nx+1, params.ny+1)
+
+        # Compute decomposition of height values at stations times the inverse covariance matrix
+        # The argument corresponds to mu10 and the outcome to mu11 in Alex's code
+        WΛWH_decomposition!(offline_matrices.buf1, mu10_sparse, offline_matrices, params)
+
+        # Compute the mean for the ith particle using mu2 and mu11
+        # Skip storing the temporary mu1 in Alex's code
+        mean[iprt,:,:] .= @view(height[iprt,:,:]) .- offline_matrices.buf1 .+ offline_matrices.buf2
+
+    end
+
+end
+
+function sample_height_proposal!(samples::AbstractArray{T,3}, height::AbstractArray{T,3}, mean::AbstractArray{T,3},
+                                 offline_matrices::OfflineMatrices, online_matrices::OnlineMatrices,
+                                 observations::AbstractVector{T}, stations::StationVectors, params::TestParameters,
+                                 rng::Random.AbstractRNG) where {T,S}
+
+    @assert params.nprt % 2 == 0 "Number of particles must be even"
+
+    calculate_mean_height!(mean, height, offline_matrices, observations, stations, params)
+
+    #TODO: Can you create a Normal(0,1) distribution of complex type?
+    nd = Normal(0,1)
+
+    i_n1 = LinearIndices((params.nx+1, params.ny+1))
+    i_n1_bar = LinearIndices((2*params.nx, 2*params.ny))
+
+    for iprt in 1:2:params.nprt
+        # TODO randn(ComplexF64) seems to set variance = 1/2. Could not find how to change that.
```

Comment:
I'm not sure what you mean:
```julia
julia> using Statistics

julia> M = randn(ComplexF64, (100, 100));

julia> std(M)
1.0015637176495376

julia> mean(M)
-0.005198967157672335 - 0.0066281896290166595im
```
The mean is about 0 and the standard deviation is about 1, what's the issue?

Reply:
Hmm, ok, I didn't actually test that. I trusted what it said here https://docs.julialang.org/en/v1/stdlib/Random/#Base.randn

Is the difference whether the variance of the complex number is 1 or the variances of the real and imaginary parts are 1? In Alex's code it was the latter.

Reply:
The documentation is correct.  This is how complex `randn` is implemented: https://github.com/JuliaLang/julia/blob/a7848a28e5d450dcc48ef99320d3220618dba501/stdlib/Random/src/normal.jl#L86-L91.  Having the real and the imaginary parts with standard deviation 1, individually, means that the complex numbers have standard deviation `sqrt(2)`:
```julia
julia> using Statistics

julia> M = complex.(randn(100, 100), randn(100, 100));

julia> std(M)
1.4117042745940547

julia> mean(M)
-0.006128621492362236 - 0.011696050838561518im
```
Looking at the paper, I'm mildly confident we want to have complex numbers with standard deviation 1, not `sqrt(2)`